aggregator/entry.go

```diff
@@ -580,6 +613,136 @@ func (e *Entry) addUntimedWithLock(timestamp time.Time, mu unaggregated.MetricUn
+	// within the time lock. This ensures time ordering by wrapping
+	// actions that need to happen before a given time within a read lock,
+	// so it is guaranteed that actions before when a write lock is acquired
+	// must have all completed. This is used to ensure we never write metrics
+	// for times that have already been flushed.
+	currTime := e.opts.ClockOptions().NowFn()()
+	e.recordLastAccessed(currTime)
+
+	e.RLock()
+	if e.closed {
+		e.RUnlock()
+		timeLock.RUnlock()
+		return errEntryClosed
+	}
+
+	// Reject datapoints that arrive too late or too early.
+	if err := e.checkTimestampForTimedMetric(
+		currTime.UnixNano(),
+		metric.TimeNanos,
+		metadata.StoragePolicy.Resolution().Window,
+	); err != nil {
+		e.RUnlock()
+		timeLock.RUnlock()
+		return err
+	}
+
+	// Check if we should update metadata, and add metric if not.
+	key := aggregationKey{
+		aggregationID:  metadata.AggregationID,
+		storagePolicy:  metadata.StoragePolicy,
+		idMutationType: IDMutationDisabled,
+	}
+	if idx := e.aggregations.index(key); idx >= 0 {
+		err := e.addTimedWithLock(e.aggregations[idx], metric)
+		e.RUnlock()
+		timeLock.RUnlock()
+		return err
+	}
+	e.RUnlock()
+
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		timeLock.RUnlock()
+		return errEntryClosed
+	}
+
+	if idx := e.aggregations.index(key); idx >= 0 {
+		err := e.addTimedWithLock(e.aggregations[idx], metric)
+		e.Unlock()
+		timeLock.RUnlock()
+		return err
+	}
+
+	// Update metatadata if not exists, and add metric.
+	if err := e.updateTimedMetadataWithLock(metric, metadata); err != nil {
+		e.Unlock()
+		timeLock.RUnlock()
+		return err
+	}
+	idx := e.aggregations.index(key)
+	err := e.addTimedWithLock(e.aggregations[idx], metric)
+	e.Unlock()
+	timeLock.RUnlock()
+	return err
+}
+
+func (e *Entry) checkTimestampForTimedMetric(
+	currNanos, metricTimeNanos int64,
+	resolution time.Duration,
+) error {
+	bufferFutureFn := e.opts.BufferForFutureTimedMetricFn()
+	timedBufferFuture := bufferFutureFn()
+	if metricTimeNanos-currNanos > timedBufferFuture.Nanoseconds() {
+		e.metrics.timed.arrivedTooEarly.Inc(1)
+		return errArrivedTooEarly
+	}
+	bufferPastFn := e.opts.BufferForPastTimedMetricFn()
+	timedBufferPast := bufferPastFn(resolution)
+	if currNanos-metricTimeNanos > timedBufferPast.Nanoseconds() {
+		e.metrics.timed.arrivedTooLate.Inc(1)
+		return errArrivedTooLate
+	}
+	return nil
+}
+
+func (e *Entry) updateTimedMetadataWithLock(
+	metric aggregated.Metric,
+	metadata metadata.TimedMetadata,
+) error {
+	var (
+		elemID = e.maybeCopyIDWithLock(metric.ID)
+		err    error
+	)
+
+	// Update the timed metadata.
+	key := aggregationKey{
+		aggregationID:  metadata.AggregationID,
+		storagePolicy:  metadata.StoragePolicy,
+		idMutationType: IDMutationDisabled,
+	}
+	listID := timedMetricListID{
+		resolution: metadata.StoragePolicy.Resolution().Window,
+	}.toMetricListID()
+	newAggregations, err := e.addNewAggregationKey(metric.Type, elemID, key, listID, e.aggregations)
```

Comment:
NIT: not part of this diff but it seems like this function should be `addNewAggregatedKeyWithLock` to conform to norms.

Reply:
Sounds good